ユーザー管理機能

YYprogramer opened the pull request:
# What
ユーザー管理機能のコーディング

# Why
ユーザー管理機能を実装するため


必要な情報を適切に入力して「会員登録」ボタンを押すと、ユーザーの新規登録ができる動画
https://gyazo.com/4d7339fd3553246e88432ce2f2c5e6a3

 入力に問題のある状態で「会員登録」ボタンが押された場合、ユーザーの新規登録はできず、エラーメッセージが出力される動画
https://gyazo.com/5b42e43b3b75dd315cdf198bd6448490

 必要な情報を適切に入力して「ログイン」ボタンを押すと、ユーザーのログインができる動画
https://gyazo.com/e7881c8c4ff4328505890a8ad5180f7c

 入力に問題のある状態で「ログイン」ボタンが押された場合、ユーザーのログインはできず、エラーメッセージが出力される動画
https://gyazo.com/f6b7be14afd6361f30a22a5ef3684623

 ログアウトに成功する動画
https://gyazo.com/9917f04f0eab3e7fdf37421252afa27e

 テストに成功する画像
https://gyazo.com/f8c7256d982f38ce398d3cb4c9fce3ec

YYprogramer committed:
Userテーブルの作成

YYprogramer committed:
ユーザー管理機能の編集

新規登録・ログインのビューファイルへのパスを繋ぐ

YYprogramer committed:
ユーザー情報の登録

YYprogramer committed:
ユーザー情報の登録

パスワード設定時に６文字以上半角英数字のバリデーションを設定しています

YYprogramer committed:
トップページでのリダイレクト

未ログインユーザーがインデックスもしくはショウアクション以外を処理しようとした場合、インデックスアクションが呼び出され、トップページへ遷移する。

YYprogramer committed:
パスワードの入力

パスワードは半角英数字のみ入力を許可。１つ以上の英字、１つ以上の数字を含むことを確認。英字と数字の組み合わせを許可。

YYprogramer committed:
メールアドレスの入力

メールアドレスは空欄では保存できない。
英字数字特殊記号＋＠＋英字数字特殊記号の文字列でなければ保存できない

YYprogramer committed:
ニックネーム

ニックネームが空欄だと保存できない

YYprogramer committed:
名前入力欄

漢字の姓名
カナの姓名
それぞれ空欄では保存できない

YYprogramer committed:
誕生日欄

誕生日欄が空欄だと保存できない

YYprogramer committed:
名前入力欄（漢字）

漢字・ひらがな・カタカナの入力のみ許可

YYprogramer committed:
名前入力欄（カナ）

カタカナの入力のみ許可

YYprogramer committed:
ログアウト機能の実装

YYprogramer committed:
インデックスページに遷移しないエラーを解消

YYprogramer committed:
未ログインユーザーのアクション修正

未ログインユーザーでも新規登録・ログインにはアクセスできるように修正

YYprogramer committed:
メールアドレス登録の修正

正規表現に不具合あり。
修正

YYprogramer committed:
オートフォーカスの修正

メールアドレス入力欄からニックネーム入力欄にオートフォーカスの修正

YYprogramer committed:
ログアウト機能の実装

YYprogramer committed:
ログイン機能の実装

YYprogramer committed:
Rspecの導入

YYprogramer committed:
単体テストコードの実装

YYprogramer committed:
エラーメッセージの表示

YYprogramer committed:
ファクトリーボットの導入と単体テストコードの追加

YYprogramer committed:
単体テストコードの簡略化

ファクトリーボットを使用して、コードの簡略化を行なった

YYprogramer committed:
ファクトリーボットの生成

YYprogramer committed:
単体テストコードの追加

YYprogramer committed:
ログイン画面にSigned out successfully.を表示させない

ログアウト直後にログイン画面に遷移すると
Signed out successfully.
のフラッシュメーセージが表示される。
そのため、トップページに非表示状態でフラッシュメッセージ表示処理を一度行うことで、ログイン時に表示されないようにしている。

yuri240 commented on app/models/user.rb:
```diff
@@ -0,0 +1,17 @@
+
+
+  validates :nickname, presence: true
+  validates :password, presence: true, length: { minimum: 6 }, format: { with: /\A(?=.*?[a-zA-Z])(?=.*?\d)/ },confirmation: true
+  validates :password_confirmation, presence: true
+  validates :email, presence: true, format: { with: /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i }
```
```ruby
  validates :password_confirmation, presence: true
  validates :email, presence: true, format: { with: /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i }
```

◎password_confirmationとemailに対するバリデーションを削除しましょう。

■理由
deviseの機能によりデフォルトで設定されるためです。

YYprogramer replied:
了解しました。
emailとpassword_confirmationに対するバリデーションは削除します。

yuri240 commented on app/models/user.rb:
```diff
@@ -0,0 +1,17 @@
+
+
+  validates :nickname, presence: true
+  validates :password, presence: true, length: { minimum: 6 }, format: { with: /\A(?=.*?[a-zA-Z])(?=.*?\d)/ },confirmation: true
```
```ruby
validates :password, presence: true, length: { minimum: 6 }, format: { with: /\A(?=.*?[a-zA-Z])(?=.*?\d)/ },confirmation: true
```
passwordに対するpresence: trueは不要となりますので、削除しておきましょう！

■理由
deviseの機能によりデフォルトで設定されるためです。

YYprogramer replied:
修正しました

yuri240 commented on app/models/user.rb:
```diff
@@ -0,0 +1,17 @@
+  validates :password, presence: true, length: { minimum: 6 }, format: { with: /\A(?=.*?[a-zA-Z])(?=.*?\d)/ },confirmation: true
+  validates :password_confirmation, presence: true
+  validates :email, presence: true, format: { with: /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i }
+  validates :family_name_kanji, presence: true, format: { with: /\A[一-龯ぁ-んァ-ン]+\z/ }
+  validates :given_name_kanji, presence: true, format: { with: /\A[一-龯ぁ-んァ-ン]+\z/ }
+  validates :family_name_kana, presence: true, format: { with: /\A[ァ-ン]+\z/ }
```
◎family_name_kanjiとgiven_name_kanjiの正規表現について
```
/\A[一-龯ぁ-んァ-ン]+\z/
```
こちら「佐々木」などである「々」が入力された際に、
弾かれないよう正規表現を下記のように修正しましょう。
↓
```
/\A[ぁ-んァ-ヶ一-龥々ー]+\z/
```

YYprogramer replied:
了解しました
修正します

yuri240 commented on app/models/user.rb:
```diff
@@ -0,0 +1,17 @@
+  validates :family_name_kanji, presence: true, format: { with: /\A[一-龯ぁ-んァ-ン]+\z/ }
+  validates :given_name_kanji, presence: true, format: { with: /\A[一-龯ぁ-んァ-ン]+\z/ }
+  validates :family_name_kana, presence: true, format: { with: /\A[ァ-ン]+\z/ }
+  validates :given_name_kana, presence: true, format: { with: /\A[ァ-ン]+\z/ }
```
◎family_name_kanaとgiven_name_kanaの正規表現について
```ruby
/\A[ァ-ン]+\z/ 
```
こちらの正規表現ですと、全角カタカナの長音符（ー）や「ヶ」を含む文字が登録できないかと思いますので、以下に修正しましょう！
↓
```ruby
/\A[ァ-ヶー]+\z/
```

YYprogramer replied:
了解しました。
修正します。

yuri240 commented on app/models/user.rb:
```diff
@@ -0,0 +1,17 @@
+
+
+  validates :nickname, presence: true
+  validates :password, presence: true, length: { minimum: 6 }, format: { with: /\A(?=.*?[a-zA-Z])(?=.*?\d)/ },confirmation: true
```
◎passwordの正規表現について

```ruby
/\A(?=.*?[a-zA-Z])(?=.*?\d)/
```
上記は、特殊文字（!）や空白、他の記号等が含まれている場合、登録できてしまうかと思いますので、以下に修正しましょう！
↓
```ruby
\A(?=.*?[a-z])(?=.*?[\d])[a-z\d]+\z/i
```

YYprogramer replied:
了解しました。
修正します。

yuri240 commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,82 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+  before do
```
正常系の「`登録できること`」というテストが不足しているようなので追加しましょう！

またその際は、
**context**を用いて「`登録できる場合`」と「`登録できない場合`」
にまとめていただくとより効率的で見やすいコードとなります。

よろしければ、下記カリキュラムを参考にしてください。
https://master.tech-camp.in/v2/curriculums/4795

YYprogramer replied:
承知しました。
正常系のテストコードを追加します

yuri240 reviewed (commented):
レビュー依頼ありがとうございます！
コメントさせていただきましたので、ご確認お願いいたします。

また、お手数ではありますが
テストコードを追加した後は、改めてテストコード実行後のキャプチャ画像を撮影いただき、
こちらのプルリクエストへ添付していただくようお願いいたします。

YYprogramer committed:
バリデーションの削除

email及びpassword_confirmationはdeviceのデフォルト機能でバリデーションが設定されているため、削除しました。

YYprogramer committed:
passwordのバリデーション修正

パスワードにおけるpresence:trueはdeviceのデフォルト機能で備わっているため、削除しました

YYprogramer committed:
正規表現の修正

〜name_kanjiにおける正規表現を
/\A[一-龯ぁ-んァ-ン]+\z/
から
/\A[ぁ-んァ-ヶ一-龥々ー]+\z/
に修正。
当初の表現だと「々」が弾かれるため

YYprogramer committed:
正規表現の修正

〜name_kanaも
/\A[ァ-ン]+\z/
から
/\A[ァ-ヶー]+\z/
に修正。
当初の表現だと「ー」や「ヶ」が入力できない

YYprogramer committed:
正規表現の修正

パスワードの正規表現を
/\A(?=.*?[a-zA-Z])(?=.*?\d)/
から
/\A(?=.*?[a-z])(?=.*?[\d])[a-z\d]+\z/i
に修正。

先頭から末尾までの文字列（\A は行の先頭を表す）
少なくとも1つのアルファベット（小文字）を含む（(?=.*?[a-z])）
少なくとも1つの数字を含む（(?=.*?[\d])）
アルファベット（小文字）と数字のみで構成される（[a-z\d]+）
以上の条件を全て満たす（\z は行の末尾を表す）

YYprogramer committed:
テストコードの修正

・正常系テストの追加
・うまくいく場合といかない場合のcontextを作成し、見やすくしている

YYprogramer committed:
テストコードの修正

不要な空白を削除

YYprogramer commented:
#What
ユーザー管理機能実装にためにコードレビューで指摘があった箇所を修正
#Why
ユーザー管理機能を実装するため

テストに成功する動画
https://gyazo.com/1ec63380f55fb7c0f94df3aa9fa5494d

TAKADA-yuma commented on app/controllers/furimas_controller.rb:
```diff
@@ -2,4 +2,9 @@ class FurimasController < ApplicationController
     render 'items/index'
   end
+
+  def destroy
+    session[:user_id] = nil
+    redirect_to root_path
+  end
```
6〜9行目
これらの記述は今後の実装で追加予定のためコメントアウトしておきましょう。

YYprogramer replied:
承知しました。
コメントアウトしておきます。

TAKADA-yuma commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,110 @@
+        @user.given_name_kana = '亜'
+        @user.valid?
+        expect(@user.errors.full_messages).to include("Given name kana is invalid")
+      end
```
以下の異常系テストも実施しましょう。
```
・姓（全角）が空だと登録できない
・姓（全角）に半角文字が含まれていると登録できない
・名（全角）が空だと登録できない
・名（全角）に半角文字が含まれていると登録できない
・姓（カナ）が空だと登録できない
・名（カナ）が空だと登録できない
```

YYprogramer replied:
了解しました
異常系のテストコードを追加します

TAKADA-yuma reviewed (commented):
レビュー依頼ありがとうございます。プルリクエスト確認させていただきました。

修正点を追加でコメントさせて頂きましたので、ご確認をお願い致します。

```ruby
【次回プルリクエスト提出する際の注意点】
①修正が終わりましたら、`コミットとプッシュ`をして、`プルリクエストに修正内容が反映されているか確認`しましょう。
②既に頂いている挙動確認の画像や動画が修正範囲だった場合、最新の画像や動画の提出をお願い致します。
③googleフォームの送信をお願いいたします。
```


ご不明点や疑問点ございましたら、チャット・コネクトメンターにご相談ください！
どうぞよろしくおねがいします。

YYprogramer committed:
ユーザーコントローラーの修正

destroyアクションはまだ実装しないためコメントアウト

YYprogramer committed:
異常系テストコードの追加

YYprogramer commented:
#What
ユーザー管理機能実装にためにコードレビューで指摘があった箇所を修正
#Why
ユーザー管理機能を実装するため

テストに成功する動画
https://gyazo.com/80c74643fc02d7fcb3b3f8b52cccf66f

koyamatakumi commented on app/models/user.rb:
```diff
@@ -0,0 +1,15 @@
+
+
+  validates :nickname, presence: true
+  validates :password, length: { minimum: 6 }, format: { with: /\A(?=.*?[a-z])(?=.*?[\d])[a-z\d]+\z/i },confirmation: true
```
```ruby
  validates :password, length: { minimum: 6 }, format: { with: /\A(?=.*?[a-z])(?=.*?[\d])[a-z\d]+\z/i },confirmation: true
```
 length: { minimum: 6 }と,confirmation: trueはデフォルトで設定されているため記述する必要はありません。

YYprogramer replied:
承知しました

koyamatakumi reviewed (commented):
YYprogramer様
レビュー依頼フォームの送信ありがとうございます。
修正いただいた内容を確認させていただき、追加でコメントいたしましたのでご確認お願いいたします。

修正後はコミット、プッシュを忘れずに行い、改めてレビュー依頼フォームの送信をお願い致します！

着実にゴールに近づいています。ラストスパート頑張っていきましょう！

YYprogramer committed:
バリデーションの修正

パスワードにおけるバリデーションで
length: { minimum: 6 }と,confirmation: trueはデフォルトで設定されているため記述を削除

YYprogramer commented:
#What
ユーザー管理機能実装にためにコードレビューで指摘があった箇所を修正
#Why
ユーザー管理機能を実装するため

// LGTMもらえますように🙏

wad-sho commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,141 @@
+        @user.password = '123456'
+        @user.valid?
+        expect(@user.errors.full_messages).to include("Password is invalid")
+      end
```
passwordに全角文字を含むと登録できないテストを追記いたしましょう！

YYprogramer replied:
追加しました

wad-sho reviewed (commented):
実装お疲れ様です！レビューのご依頼ありがとうございます。
コメントいたしましたので、ご確認頂けますと幸いでございます。

ご修正後再度レビュー依頼のほどよろしくお願いいたします！

YYprogramer commented:
#What
ユーザー管理機能実装にためにコードレビューで指摘があった箇所を修正
#Why
ユーザー管理機能を実装するため

テストに成功する画像
tps://gyazo.com/30037d91394f71625d861f8c665a314a

AsukaToyoshima reviewed (commented):
レビューのご依頼ありがとうございます。
こちらのプルリクエストですが、コードの変更が確認できませんでした。
GitHubDesktopでコミットとプッシュができているか一度ご確認いただき、再度レビュー依頼をいただけたらと存じます。
疑問点・不明点等ございましたら、コネクトメンターへお気軽にご質問ください。

YYprogramer committed:
テストコードの追加

パスワードに全角文字が含まれると保存できないテストコードを追加

YYprogramer commented:
#What
ユーザー管理機能実装にためにコードレビューで指摘があった箇所を修正
#Why
ユーザー管理機能を実装するため

テストに成功する画像
tps://gyazo.com/30037d91394f71625d861f8c665a314a

sh10n7 reviewed (commented):
修正とレビュー依頼ありがとうございます。

修正点確認いたしました。
しっかりと実装できていることが確認できましたので、**LGTM**とさせていただきます！
実装お疲れ様でした。

こちらのブランチは忘れずにMergeを行っていただき、ローカルのmasterブランチにPullしましょう。

その後、新しいブランチを作成し次の実装に取り組んでいただけますと幸いです。
ここまでの実装、本当にお疲れ様でした。
引き続き、応援しております！
